# Gates and circuit construction

[`01_first_circuit.ipynb`](01_first_circuit.ipynb) used two gates, H and CX, to build the Bell state. Polypus offers a wider set of gates, and this notebook goes through them, along with a few more ways to build circuits.

## Single-qubit gates

| Gate | What it does |
|---|---|
| `h(q)` | Puts the qubit in superposition, as seen in [`00b`](00b_quantum_intuition.ipynb) and in [`01_first_circuit.ipynb`](01_first_circuit.ipynb). |
| `x(q)` | Flips 0 and 1: the quantum NOT. |
| `y(q)` | Like `x`, and also changes the state's phase. |
| `z(q)` | Changes the state's phase without touching the value measured in the usual basis. Its effect is only noticeable combined with superposition: on a qubit already in a definite value, a measurement cannot tell whether it was applied or not. |
| `s(q)`, `sdg(q)` | Quarter-turn phase rotations, one in each direction. |
| `t(q)`, `tdg(q)` | Eighth-turn phase rotations. T comes back later in the series, when decomposing more complex gates. |
| `rx(q, theta)`, `ry(q, theta)`, `rz(q, theta)` | Continuous rotations around each axis, by a chosen angle `theta`. |
| `u(q, theta, phi, lam)` | The most general single-qubit gate: any single-qubit operation can be expressed this way. |

`x`, `y`, and `z` are, at bottom, specific cases of the rotations `rx`/`ry`/`rz`, with the angle fixed to π: the difference with the latter is that their angle is fixed instead of chosen. `h` does not belong to this group: it is not a rotation around one of those three axes, but a combination of `x` and `z`.

Unlike `h`, which always splits the probabilities 50/50, a rotation like `rx` allows choosing the proportion:

In [1]:
import math

import polypus

qc = polypus.Circuit(1).rx(0, math.pi / 3).measure_all()
result = polypus.run_quantum_circuit(qc, shots=1000, infrastructure="local")
print(result.counts[0])

{'0': 747, '1': 253}


The result comes close to 25% `'1'` and 75% `'0'`, instead of the 50/50 of `h`. Changing the angle changes that proportion: this is the basis for the trainable circuits that appear in [`06a_variational_training.ipynb`](06a_variational_training.ipynb).

## Two-qubit gates

| Gate | What it does |
|---|---|
| `cx(q0, q1)` | CNOT: flips `q1` if `q0` is 1, as seen in [`01_first_circuit.ipynb`](01_first_circuit.ipynb). |
| `cz(q0, q1)` | Like `cx`, but changes the phase instead of the value, and is symmetric: there is no real distinction between control and target. |
| `swap(q0, q1)` | Swaps the state of the two qubits. |
| `rzz(q0, q1, theta)`, `rxx(q0, q1, theta)` | Two-qubit rotations, with a chosen angle. Heavily used in the trainable circuits of [`06a_variational_training.ipynb`](06a_variational_training.ipynb). |
| `cp(q0, q1, theta)` | Controlled phase: like `cz`, but with a chosen angle instead of a fixed one. Comes back in the quantum Fourier transform, [`04b_qft.ipynb`](04b_qft.ipynb). |

## Other operations

- `barrier()`: does not change the circuit's state, it only marks a visual or organizational separation between groups of gates.
- `measure(qubit, cbit)`: measures a single specific qubit, unlike `measure_all()`, which measures all of them at once.

The two tables above are the gates the rest of the series uses, not the full list: Polypus also has controlled versions of several of them (`ch`, `crx`, `cry`, `crz`, `cu`...), multi-controlled gates like `ccx` or `c3x`, used in [`05b_grover.ipynb`](05b_grover.ipynb), and a few gates inherited from Qiskit under their own name (`p`, `u1`, `u2`, `sx`...). The full list is in `dir(polypus.Circuit)`.

## One more example: the GHZ state

The same pattern from the Bell state, an H followed by CX, extends naturally to more qubits. With three qubits and two CX from the same control, the so-called **GHZ state** is obtained:

In [2]:
ghz = polypus.Circuit(3).h(0).cx(0, 1).cx(0, 2).measure_all()
result = polypus.run_quantum_circuit(ghz, shots=1000, infrastructure="local")
print(result.counts[0])

{'000': 506, '111': 494}


Just like with the Bell state, only `'000'` and `'111'` show up, never a mix: all three qubits end up entangled at once.

## Summary

This notebook went through the rest of Polypus's gates, single-qubit and two-qubit, with two examples run live: a rotation with adjustable probability, and an extension of the Bell state to three qubits.

[`02b_qiskit_qasm_interop.ipynb`](02b_qiskit_qasm_interop.ipynb) covers how these circuits interoperate with Qiskit and with OpenQASM.

## Next step

Continue with: [`02b_qiskit_qasm_interop.ipynb`](02b_qiskit_qasm_interop.ipynb).